internal/envoy/envoy.go

```diff
@@ -40,53 +43,133 @@ const (
 	configFileName       = "envoy-config.yaml"
 )
 
+type serverOptions struct {
+	services       string
+	logLevel       string
+	tracingOptions trace.TracingOptions
+}
+
 // A Server is a pomerium proxy implemented via envoy.
 type Server struct {
 	wd  string
 	cmd *exec.Cmd
 
 	grpcPort, httpPort string
-	opts               *config.Options
+	envoyPath          string
+	restartEpoch       int
+
+	mu      sync.Mutex
+	options serverOptions
 }
 
 // NewServer creates a new server with traffic routed by envoy.
-func NewServer(opts *config.Options, grpcPort, httpPort string) (*Server, error) {
+func NewServer(src config.Source, grpcPort, httpPort string) (*Server, error) {
 	wd := filepath.Join(os.TempDir(), workingDirectoryName)
 	err := os.MkdirAll(wd, 0755)
 	if err != nil {
 		return nil, fmt.Errorf("error creating temporary working directory for envoy: %w", err)
 	}
 
-	srv := &Server{
-		wd:       wd,
-		grpcPort: grpcPort,
-		httpPort: httpPort,
-		opts:     opts,
+	envoyPath, err := extractEmbeddedEnvoy()
+	if err != nil {
+		log.Warn().Err(err).Send()
+		envoyPath = "envoy"
 	}
 
-	err = srv.writeConfig()
-	if err != nil {
-		return nil, fmt.Errorf("error writing initial envoy configuration: %w", err)
+	srv := &Server{
+		wd:        wd,
+		grpcPort:  grpcPort,
+		httpPort:  httpPort,
+		envoyPath: envoyPath,
 	}
 
+	src.OnConfigChange(srv.onConfigChange)
+	srv.onConfigChange(src.GetConfig())
+
 	return srv, nil
 }
 
-// Run runs the server by extracting the embedded envoy and then executing it.
-func (srv *Server) Run(ctx context.Context) error {
-	envoyPath, err := extractEmbeddedEnvoy()
+// Close kills any underlying envoy process.
+func (srv *Server) Close() error {
+	srv.mu.Lock()
+	defer srv.mu.Unlock()
+
+	var err error
+	if srv.cmd != nil && srv.cmd.Process != nil {
+		err = srv.cmd.Process.Kill()
+		if err != nil {
+			log.Error().Err(err).Str("service", "envoy").Msg("envoy: failed to kill process on close")
+		}
+		srv.cmd = nil
+	}
+
+	return err
+}
+
+func (srv *Server) onConfigChange(cfg *config.Config) {
+	srv.update(cfg)
+}
+
+func (srv *Server) update(cfg *config.Config) {
+	srv.mu.Lock()
+	defer srv.mu.Unlock()
+
+	tracingOptions, err := config.NewTracingOptions(cfg.Options)
 	if err != nil {
-		log.Warn().Err(err).Send()
-		envoyPath = "envoy"
+		log.Error().Err(err).Str("service", "envoy").Msg("invalid tracing config")
+		return
 	}
 
-	srv.cmd = exec.CommandContext(ctx, envoyPath,
+	options := serverOptions{
+		services:       cfg.Options.Services,
+		logLevel:       firstNonEmpty(cfg.Options.ProxyLogLevel, cfg.Options.LogLevel, "debug"),
+		tracingOptions: *tracingOptions,
+	}
+
+	if cmp.Equal(srv.options, options, cmp.AllowUnexported(serverOptions{})) {
+		log.Debug().Str("service", "envoy").Msg("envoy: no config changes detected")
+		return
+	}
+	srv.options = options
+
+	// release the previous process so we can hot-reload
```

Comment:
Can/should we do this after a successful spawn of the new process?

Reply:
I think this is a bit tricky, cause there's not an easy way for us to tell when the old process is ready to be killed. 

The danger would be that we don't kill the child process... but then if pomerium is restarted, I think it will pick up the envoy that was already running, because the temporary base id file still exists. So maybe this is mostly a non-issue?

Reply:
Just to make sure I'm clear on the behavior here...envoy self terminates the old process, right?  The `Release()` call just cleans up our references to the running process and doesn't actively kill it.  I didn't think we had to tell the old envoy process when to be killed.

Assuming that's all true - if we weren't able to start a new envoy,  we'd continue running with the old envoy (granted, without a reference to it) until a new one starts successfully.  That seems okay.

However, if we were in that state and received a kill signal, wouldn't we want to terminate the original envoy?  I don't think we'd be able to with this order of operations.

You're probably right that in a restart scenario, the abandoned process would get taken over with a new envoy if it comes up.  However, I think it's reasonable and a little less surprising if envoy terminates when the pomerium process does.

Thoughts?  There might be something I'm missing

Reply:
I moved the process release to after we start the new process.